### PR TITLE
Block evictions

### PR DESCRIPTION
**Pull request by minghinmatthewlam**
- Configurable value for minimum age a candidate must pass before allowed eviction
- When a candidate is nominated for eviction but does not pass the minimum age, we trace information to figure out why.
- - Eviction attempts before the candidate
- Eviction candidate replica count, effective age, and age
- Pool size, window size, removal fraction from config

### CODE REVIEW

File: Public/Src/Cache/ContentStore/Distributed/Redis/RedisContentLocationStoreConfiguration.cs [hunk 1]
```diff
@@ -92,6 +92,11 @@ public class RedisContentLocationStoreConfiguration : LocalLocationStoreConfigur
         /// </remarks>
         public int EvictionPoolSize { get; set; } = 5000;
 
+        /// <summary>
+        /// 
+        /// </summary>
+        public TimeSpan EvictionMinAge { get; set; } = TimeSpan.FromMinutes(30);
```
Comment (SergeyTeplyakov) [inline Public/Src/Cache/ContentStore/Distributed/Redis/RedisContentLocationStoreConfiguration.cs]: Could you please add a comment, or use <nodoc />?

File: Public/Src/Cache/DistributedCache.Host/Configuration/DistributedContentSettings.cs [hunk 2]
```diff
@@ -215,6 +215,12 @@ private DistributedContentSettings(IDictionary<string, string> connectionSecretN
         [DataMember]
         public int EvictionPoolSize { get; set; } = 5000;
 
+        /// <summary>
+        /// A candidate must have an age older than this amount, or else it won't be evicted.
+        /// </summary>
+        [DataMember]
+        public TimeSpan EvictionMinAge { get; set; } = TimeSpan.FromMinutes(30);
```
Comment (SergeyTeplyakov) [inline Public/Src/Cache/DistributedCache.Host/Configuration/DistributedContentSettings.cs]: The feature should be disabled once it checked in.

It means that by default the time should be 0. Right?

File: Public/Src/Cache/ContentStore/Test/Stores/FileSystemContentStoreInternalTests.cs [hunk 3]
```diff
@@ -259,6 +259,7 @@ private static DistributedEvictionSettings WithMockDistributedLocationStore(Mock
                 (context, contentHashesWithInfo, cts, urgencyHint) => null,
                 locationStoreBatchSize: 42,
                 replicaCreditInMinutes: null,
+                minAgeBeforeEviction: null,
```
Comment (SergeyTeplyakov) [inline Public/Src/Cache/ContentStore/Test/Stores/FileSystemContentStoreInternalTests.cs]: Is it still needed?

File: Public/Src/Cache/ContentStore/Distributed/NuCache/TransitioningContentLocationStore.cs [hunk 4]
```diff
@@ -313,7 +315,22 @@ public Task<BoolResult> UnregisterAsync(Context context, IReadOnlyList<ContentHa
...
+
+        private bool IsPassEvictionAge(Context context, ContentHashWithLastAccessTimeAndReplicaCount candidate, TimeSpan evictionMinAge, ref int evictionCount)
+        {
+            if (candidate.Age > evictionMinAge)
+            {
+                evictionCount++;
+                return true;
+            }
+
+            context.Debug($"{evictionCount} eviction attempts before current candidate");
```
Comment (SergeyTeplyakov) [inline Public/Src/Cache/ContentStore/Distributed/NuCache/TransitioningContentLocationStore.cs]: This should be traced in one line.

And I believe that the third line you have right now is pretty stable and I don't think we need to trace it all the time.

File: Public/Src/Cache/ContentStore/Distributed/Redis/RedisContentLocationStoreConfiguration.cs (hunk 1 above)
Comment (minghinmatthewlam) [reply Public/Src/Cache/ContentStore/Distributed/Redis/RedisContentLocationStoreConfiguration.cs]: added

File: Public/Src/Cache/DistributedCache.Host/Configuration/DistributedContentSettings.cs (hunk 2 above)
Comment (minghinmatthewlam) [reply Public/Src/Cache/DistributedCache.Host/Configuration/DistributedContentSettings.cs]: changed default to TimeSpan.Zero, configure value when we need to test for it.

File: Public/Src/Cache/ContentStore/Test/Stores/FileSystemContentStoreInternalTests.cs (hunk 3 above)
Comment (minghinmatthewlam) [reply Public/Src/Cache/ContentStore/Test/Stores/FileSystemContentStoreInternalTests.cs]: removed

File: Public/Src/Cache/ContentStore/Distributed/NuCache/TransitioningContentLocationStore.cs [hunk 5]
```diff
@@ -27,23 +28,28 @@ namespace BuildXL.Cache.ContentStore.Distributed.NuCache
     /// <summary>
     /// <see cref="IContentLocationStore"/> implementation that supports old redis and new local location store.
     /// </summary>
-    internal class TransitioningContentLocationStore : StartupShutdownBase, IContentLocationStore, IDistributedLocationStore
+    internal class
+        TransitioningContentLocationStore : StartupShutdownBase, IContentLocationStore, IDistributedLocationStore
```
Comment (SergeyTeplyakov) [inline Public/Src/Cache/ContentStore/Distributed/NuCache/TransitioningContentLocationStore.cs]: nit: can you please undo this formatting?

File: Public/Src/Cache/ContentStore/Distributed/NuCache/TransitioningContentLocationStore.cs [hunk 6]
```diff
@@ -313,7 +322,20 @@ public Task<BoolResult> UnregisterAsync(Context context, IReadOnlyList<ContentHa
...
+        }
+
+        private bool IsPassEvictionAge(Context context, ContentHashWithLastAccessTimeAndReplicaCount candidate, TimeSpan evictionMinAge, int index, ref int evictionCount)
+        {
+            if (candidate.Age(_clock) >= evictionMinAge)
+            {
+                evictionCount++;
+                return true;
+            }
+            context.Debug($"Previous successful eviction attempts = {evictionCount}, Total eviction attempts previously = {index}, minimum eviction age = {evictionMinAge.ToString()}, pool size = {_configuration.EvictionPoolSize}." +
```
Comment (SergeyTeplyakov) [inline Public/Src/Cache/ContentStore/Distributed/NuCache/TransitioningContentLocationStore.cs]: very nit: could you please add an extra blank line after the if block?

Comment (minghinmatthewlam) [reply Public/Src/Cache/ContentStore/Distributed/NuCache/TransitioningContentLocationStore.cs]: added

File: Public/Src/Cache/ContentStore/Distributed/NuCache/TransitioningContentLocationStore.cs (hunk 5 above)
Comment (minghinmatthewlam) [reply Public/Src/Cache/ContentStore/Distributed/NuCache/TransitioningContentLocationStore.cs]: fixed

File: Public/Src/Cache/ContentStore/Distributed/NuCache/TransitioningContentLocationStore.cs (hunk 6 above)
Comment (minghinmatthewlam) [reply Public/Src/Cache/ContentStore/Distributed/NuCache/TransitioningContentLocationStore.cs]: fixed

File: Public/Src/Cache/ContentStore/Distributed/NuCache/TransitioningContentLocationStore.cs (hunk 4 above)
Comment (minghinmatthewlam) [reply Public/Src/Cache/ContentStore/Distributed/NuCache/TransitioningContentLocationStore.cs]: fixed

**Pull request merged**